python/tvm/relay/testing/py_converter.py

```diff
@@ -0,0 +1,592 @@
+import ast
+from ast import alias, Assign, Load, Name, NameConstant, Num, Return, Store, Str
+import re
+
+import tvm
+from tvm import relay
+from tvm.relay.adt import Pattern
+from tvm.relay.backend import compile_engine
+from tvm.relay.expr import Expr, Function, GlobalVar, Var
+from tvm.relay.expr_functor import ExprFunctor
+
+OUTPUT_VAR_NAME = '_py_out'
+
+# corresponds to:
+#     import numpy
+#     import tvm
+#     from tvm import relay
+#     from tvm.relay.backend.interpreter import RefValue, TupleValue, TensorValue, ConstructorValue
+PROLOGUE = [
+    ast.Import([alias('numpy', None)]),
+    ast.Import([alias('tvm', None)]),
+    ast.ImportFrom('tvm', [alias('relay', None)], 0),
+    ast.ImportFrom('tvm.relay.backend.interpreter',
+                   [alias('RefValue', None),
+                    alias('TupleValue', None),
+                    alias('TensorValue', None),
+                    alias('ConstructorValue', None)],
+                   0)
+]
+
+class PythonConverter(ExprFunctor):
+    '''Functor for translating Relay programs into Python ASTs.'''
+
+    def __init__(self, mod, target) -> None:
+        super().__init__()
+        self.mod = mod
+        self.tgt = target
+        self.engine = compile_engine.get()
+        self.fun_no = 0
+        self.var_no = 0
+        self.var_map = {}
+
+
+    def convert(self, prog: Expr):
+        '''This method converts the passed Relay expression into a Python
+        AST object with equivalent semantics.
+
+        The Python AST can be executed using exec(); it can be turned
+        into text and inspected using astor.
+        '''
+        optimized = self.optimize(prog)
+
+        # start with conversion prelude (imports) and convert global defs
+        body = []
+        body += PROLOGUE
+        body += self.convert_module()
+
+        prog_body, extra_defs = self.visit(optimized)
+        body += extra_defs
+
+        # we finally must assign the final expression to the output var
+        # so it can be read after running EXEC
+        body.append(Assign([Name(OUTPUT_VAR_NAME, Store())], prog_body))
+
+        return ast.fix_missing_locations(ast.Module(body=body))
+
+
+    def optimize(self, prog: Expr):
+        '''Performs optimizations necessary to be able to generate code for prog.'''
+        # unwrap tuple wrappers (some op calls produce them)
+        unwrapped = prog.astuple() if isinstance(prog, relay.TupleWrapper) else prog
+        assert relay.analysis.well_formed(unwrapped)
+        mod = self.mod.from_expr(unwrapped, self.mod.functions, self.mod.type_definitions)
+
+        # necessary pass: SimplifyInference (otherwise we can't generate code for some operators)
+        # and fusion (to get primitive functions)
+        opts = relay.transform.Sequential([relay.transform.SimplifyInference(),
+                                           relay.transform.FuseOps(fuse_opt_level=0)])
+        opts(mod)
```

Comment:
Shouldn't be `mod = opts(mod)`?

Reply:
we can directly pass a `module` to the `optimize` function, right? because it looks that we already have them in the functions that call `optimize`

Reply:
It's not effectful? I'm surprised that the tests work then

Reply:
I'll look into it

Reply:
Appears that the API does require using the returned mod, but the fuse_ops pass mutates the mod so it worked anyway. Weird

Reply:
It happened that it is a `FunctionPass` which I made it effectful as there were some function passes implicitly change the Module.